fendermint/vm/interpreter/src/fvm/interpreter.rs

```diff
@@ -0,0 +1,532 @@
+            state.activity_tracker().record_block_committed(pubkey)?;
+        }
+
+        let checkpoint_outcome = self
+            .bottom_up_manager
+            .create_checkpoint_if_needed(&mut state)?;
+
+        let (power_updates, block_end_events) = if let Some(outcome) = checkpoint_outcome {
+            self.bottom_up_manager
+                .cast_validator_signatures_for_incomplete_checkpoints(
+                    outcome.checkpoint,
+                    &mut state,
+                )
+                .await?;
+            (outcome.power_updates, outcome.block_end_events)
+        } else {
+            (PowerUpdates::default(), BlockEndEvents::default())
+        };
+
+        let next_gas_market = state.finalize_gas_market()?;
+
+        if !power_updates.0.is_empty() {
+            self.top_down_manager
+                .update_voting_power_table(&power_updates)
+                .await;
+        }
+
+        let response = EndBlockResponse {
+            power_updates,
+            gas_market: next_gas_market,
+            events: block_end_events,
+        };
+        Ok(response)
+    }
+
+    async fn apply_message(
+        &self,
+        mut state: &mut FvmExecState<DB>,
+        msg: Vec<u8>,
+    ) -> Result<ApplyMessageResponse, ApplyMessageError> {
+        let chain_msg = match fvm_ipld_encoding::from_slice::<ChainMessage>(&msg) {
+            Ok(msg) => msg,
+            Err(e) => {
+                if self.reject_malformed_proposal {
+                    tracing::warn!(
+                        error = e.to_string(),
+                        "failed to decode delivered message as ChainMessage; may indicate a node issue"
+                    );
+                }
+                return Err(ApplyMessageError::InvalidMessage(e.to_string()));
+            }
+        };
+
+        match chain_msg {
+            ChainMessage::Signed(msg) => {
+                if let Err(e) = msg.verify(&state.chain_id()) {
+                    return Err(ApplyMessageError::InvalidSignature(e));
+                }
+                let applied_message = execute_signed_message(&mut state, msg.clone()).await?;
+                let domain_hash = msg.domain_hash(&state.chain_id())?;
+                Ok(ApplyMessageResponse {
+                    applied_message,
+                    domain_hash,
+                })
+            }
+            ChainMessage::Ipc(ipc_msg) => match ipc_msg {
+                IpcMessage::TopDownExec(p) => {
+                    let applied_message = self
+                        .top_down_manager
+                        .execute_topdown_msg(&mut state, p)
+                        .await?;
+                    Ok(ApplyMessageResponse {
+                        applied_message,
+                        domain_hash: None,
+                    })
+                }
+            },
+        }
+    }
+
+    async fn query(
+        &self,
+        state: FvmQueryState<DB>,
```

Comment:
this can be ReadOnly wrapped store?

Reply:
The `FvmQueryState` uses the read only block store internally. I did not change that.